### PR TITLE
Completed Jenkins configuration and job creation

### PR DESCRIPTION
**Pull request by Hontaro**
This Ansible role now does the following:
- Install Java on the server
- Install Jenkins
- Configure Jenkins global configuration (directory, port, plugins etc)
- Configure a Jenkins Job with the following attributes:
  - Repository path + branch
  - Polling/Timer strategy with frequency (Build the job when the repo changes or after a set amount of time)
  - Pre-build scripts
  - Ansible playbook build step (mandatory)
  - Post-build scripts
  - E-mail notification if build fails

Because the vault is currently not up, the SSH key for Jenkins is currently copied from my local source:
```
copy:
    src: /home/osboxes/Jenkins/jenkinsprivate
    dest: "{{ jenkins_home }}/.ssh/id_rsa"
```
The job is configured to use the jenkins user's SSH key, which is copied into $JENKINS_HOME/.ssh/
The defaults/main.yml lists all available variables. Variables that are commented out are mandatory and should be manually set.

Currently not supported: Accessing the Jenkins webinterface via HTTPS

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,30 +10,21 @@ Requires `curl` to be installed on the server.
...
-Optional boolean to define whether to allow Jenkins HTTPS access. Enables setting HTTPS port & Java keystore (jks) location and password.
-
-    jenkins_jks
-    jenkins_jks_password
-
-Location & password of Java keystore file for HTTPS. e.g. `/etc/ssl/jenkins-server.fqdn.com/jenkins-server.fqdn.com.jks`
+#### Jenkins Configuration
+| Name | Type | Required | Default | Description
+|------|------|:----------:|---------|-----------|
+|jenkins_hostname | String | ✔️ | localhost | The system hostname, usually `localhost`. This will be used to communicate with Jenkins |
```
Comment (SlothOfAnarchy) [inline README.md]: if you set a default, the var is not required.

File: README.md [hunk 2]
```diff
@@ -10,30 +10,21 @@ Requires `curl` to be installed on the server.
...
-
-    jenkins_https
-
-Optional boolean to define whether to allow Jenkins HTTPS access. Enables setting HTTPS port & Java keystore (jks) location and password.
-
-    jenkins_jks
-    jenkins_jks_password
-
-Location & password of Java keystore file for HTTPS. e.g. `/etc/ssl/jenkins-server.fqdn.com/jenkins-server.fqdn.com.jks`
+#### Jenkins Configuration
```
Comment (SlothOfAnarchy) [inline README.md]: please remove the build status on top of this readme, because it doesn't reference this repo

File: README.md [hunk 3]
```diff
@@ -10,30 +10,21 @@ Requires `curl` to be installed on the server.
...
+| jenkins_home | String | ✔️ | /var/lib/jenkins | The jenkins home directory, used for storing workspaces, artifacts etc. |
+| jenkins_http_port | int | ✔️ | 8080 | The HTTP Port for the Jenkins interface. Use `-1` to deactivate |
+| jenkins_https | boolean | | | Optional boolean to define whether to allow Jenkins HTTPS access. Enables setting HTTPS port and Java keystore location/password |
+| jenkins_https_port | int | | | The HTTPS Port for the Jenkins interface. |
+| jenkins_jks | String | | | Java keystore location for HTTPS |
+| jenkins_jks_password | String | | | Java keystore password |
+| jenkins_admin_username | String | ✔️ | | The admin username |
+| jenkins_admin_password | String | ✔️ | | The admin password |
 
+Available variables are listed below, along with default values (see `defaults/main.yml`):
```
Comment (SlothOfAnarchy) [inline README.md]: please move the vars and defaults to the table, it's cleaner and you don't have to change everything in two places.

File: defaults/main.yml [hunk 4]
```diff
@@ -19,16 +19,21 @@ jenkins_plugins: []
...
-#jenkins_log: /var/log/$NAME/$NAME.log
-#jenkins_user: $NAME
-#jenkins_group: $NAME
-#jenkins_war: /usr/share/$NAME/$NAME.war
-#jenkins_pidfile: /var/run/$NAME/$NAME.pid
+jenkins_repository_branch: master
+
+#jenkins_admin_username
+#jenkins_admin_password
+#jenkins_pre_build: String
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]: why do you want to keep commented-out entries?

File: meta/main.yml [hunk 5]
```diff
@@ -1,6 +1,5 @@
 ---
-dependencies:
-  - geerlingguy.java
+dependencies: []
```
Comment (SlothOfAnarchy) [inline meta/main.yml]: please remove the dependency from README

File: tasks/main.yml [hunk 6]
```diff
@@ -1,4 +1,18 @@
 ---
+
+# Install java
+- name: Install Java
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: please update repo cache beforehand using an extra task with `changed_when`: `false`

File: tasks/main.yml [hunk 7]
```diff
@@ -1,4 +1,18 @@
...
+  apt: 
+    name: openjdk-8-jre
+    state: present
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: does this package have the same name on all supported distributions?

File: tasks/main.yml [hunk 8]
```diff
@@ -1,4 +1,18 @@
 ---
+
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: consider splitting up your tasks to different files

File: tasks/main.yml [hunk 9]
```diff
@@ -32,6 +46,19 @@
 - name: Ensure Jenkins is started and runs on startup.
   service: name=jenkins state=started enabled=yes
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: indent

File: tasks/main.yml [hunk 10]
```diff
@@ -59,3 +86,26 @@
 
 # Update Jenkins and install configured plugins.
 - include: plugins.yml
+
+# Move ssh key to jenkins directory
+- name: Create ssh directory
+  file: path="{{ jenkins_home }}/.ssh" state=directory
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: indent

File: tasks/main.yml [hunk 11]
```diff
@@ -59,3 +86,26 @@
...
+
+- name: Copy Jenkins SSH key to his home dir
+  copy:
+    src: /home/osboxes/Jenkins/jenkinsprivate
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: `osboxes`?

File: tasks/main.yml [hunk 12]
```diff
@@ -59,3 +86,26 @@
...
+    dest: "{{ jenkins_home }}/.ssh/id_rsa"
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: why do you think I want to use a RSA key?
Also, please document this in README

File: tasks/main.yml [hunk 13]
```diff
@@ -59,3 +86,26 @@
...
+    dest: "{{ jenkins_home }}/credentials.xml"
+
+# Create job from template.
+- name: Generate config from template
+  template:
+    src: config.xml.j2
+    dest: /tmp/config.xml
+    
+- name: Create job
+  shell: java -jar {{jenkins_jar_location}} -s {{ (jenkins_http_port == -1) | ternary("https","http") }}://{{ jenkins_hostname }}{{ jenkins_port|default('') }}{{ jenkins_url_prefix | default("") }}/ create-job Job < /tmp/config.xml --username {{jenkins_admin_username}} --password {{jenkins_admin_password}}
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: is there no java module?

File: templates/basic-security.groovy.j2 [hunk 14]
```diff
@@ -1,13 +1,10 @@
-# {{ ansible_managed }}
```
Comment (SlothOfAnarchy) [inline templates/basic-security.groovy.j2]: why did you remove this line?

File: templates/config.xml.j2 [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+  <actions/>
+  <description></description>
+  <keepDependencies>false</keepDependencies>
+  <properties/>
+  <scm class="hudson.plugins.git.GitSCM" plugin="git@3.2.0">
+    <configVersion>2</configVersion>
+    <userRemoteConfigs>
+      <hudson.plugins.git.UserRemoteConfig>
+        <url>{{ jenkins_repository | mandatory }}</url>
+        <credentialsId>5ff70a7c-cef3-460d-ad91-59ed80eef4a1</credentialsId>
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: hard-coded credentials?

File: templates/config.xml.j2 [hunk 16]
```diff
@@ -0,0 +1,81 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<project>
+  <actions/>
+  <description></description>
+  <keepDependencies>false</keepDependencies>
+  <properties/>
+  <scm class="hudson.plugins.git.GitSCM" plugin="git@3.2.0">
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: hard-coded version?

File: templates/config.xml.j2 [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        <name>*/{{jenkins_repository_branch | default("master")}}</name>
+      </hudson.plugins.git.BranchSpec>
+    </branches>
+    <doGenerateSubmoduleConfigurations>false</doGenerateSubmoduleConfigurations>
+    <submoduleCfg class="list"/>
+    <extensions/>
+  </scm>
+  <canRoam>true</canRoam>
+  <disabled>false</disabled>
+  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: aren't all these vars something a user might want to configure?

File: templates/config.xml.j2 [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
+  <triggers>
+    {% if jenkins_timer_trigger is defined %}
+    <hudson.triggers.TimerTrigger>
+      <spec>{{jenkins_timer_trigger}}</spec>
+    </hudson.triggers.TimerTrigger>
+    {% endif %}
+    {%if jenkins_scm_trigger is defined %}
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: please document all vars in README

File: templates/config.xml.j2 [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    {% endif %}
+  </triggers>
+  <concurrentBuild>false</concurrentBuild>
+  <builders>
+    {% if jenkins_pre_build is defined %}
+    <hudson.tasks.Shell>
+      <command>{{ jenkins_pre_build }}</command>
+    </hudson.tasks.Shell>
+    {% endif %}
+    <org.jenkinsci.plugins.ansible.AnsiblePlaybookBuilder plugin="ansible@0.6.2">
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: hardcoded version?

File: templates/config.xml.j2 [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    </org.jenkinsci.plugins.ansible.AnsiblePlaybookBuilder>
+    {% if jenkins_post_build is defined %}
+    <hudson.tasks.Shell>
+      <command>{{ jenkins_post_build }}</command>
+    </hudson.tasks.Shell>
+    {% endif %}
+  </builders>
+  {% if jenkins_emails is defined %}
+  <publishers>
+    <hudson.tasks.Mailer plugin="mailer@1.20">
```
Comment (SlothOfAnarchy) [inline templates/config.xml.j2]: version?

File: templates/credentials.xml [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+<com.cloudbees.plugins.credentials.SystemCredentialsProvider plugin="credentials@2.1.13">
+  <domainCredentialsMap class="hudson.util.CopyOnWriteMap$Hash">
+    <entry>
+      <com.cloudbees.plugins.credentials.domains.Domain>
+        <specifications/>
+      </com.cloudbees.plugins.credentials.domains.Domain>
+      <java.util.concurrent.CopyOnWriteArrayList>
+        <com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey plugin="ssh-credentials@1.13">
+          <scope>GLOBAL</scope>
+          <id>5ff70a7c-cef3-460d-ad91-59ed80eef4a1</id>
```
Comment (SlothOfAnarchy) [inline templates/credentials.xml]: looks like a hardcoced UUID?

File: templates/basic-security.groovy.j2 (hunk 14 above)
Comment (Hontaro) [inline templates/basic-security.groovy.j2]: Ansible didn't remove that line, causing the file to break and the groovy script to not execute because of syntax errors

File: templates/config.xml.j2 (hunk 15 above)
Comment (Hontaro) [inline templates/config.xml.j2]: Only the credentialId is hardcoded, the credentials itself is the SSH key that is copied from the ansible role, which can be changed

File: defaults/main.yml (hunk 4 above)
Comment (Hontaro) [inline defaults/main.yml]: Commented-out entries are mainly for easy debugging since you can un-outcomment them and set a variable instead of looking into the README and pasting the variables with a value into the command line/default main.yml

File: tasks/main.yml (hunk 11 above)
Comment (Hontaro) [inline tasks/main.yml]: As stated, since the vault is currently not running (according to pi), I resorted to copying the private ssh key from a local source for now; It should be changed later on

File: defaults/main.yml (hunk 4 above)
Comment (SlothOfAnarchy) [inline defaults/main.yml]: not on the master branch though. you can keep them locally or in a develop branch

File: tasks/main.yml (hunk 11 above)
Comment (SlothOfAnarchy) [inline tasks/main.yml]: Please use a variable to set the key path and document it in readme

File: templates/basic-security.groovy.j2 (hunk 14 above)
Comment (SlothOfAnarchy) [inline templates/basic-security.groovy.j2]: you can comment groovy code using double slashes (http://docs.groovy-lang.org/next/html/documentation/#_comments)

File: README.md (hunk 1 above)
Comment (dasJ) [inline README.md]: Please indent this table

File: README.md [hunk 22]
```diff
@@ -95,7 +84,7 @@ Changes made to the Jenkins init script; the default set of changes set the conf
 
 ## Dependencies
 
-  - geerlingguy.java
+  - curl
```
Comment (dasJ) [inline README.md]: Why don't you install it?

File: defaults/main.yml [hunk 23]
```diff
@@ -19,16 +19,21 @@ jenkins_plugins: []
...
+#jenkins_emails: Array[String]
+#jenkins_repository: String
+#jenkins_repository_branch: String
+#jenkins_timer_trigger: Cron Syntax
+#jenkins_scm_trigger: Cron Syntax
 
 jenkins_init_changes:
   - option: "JENKINS_ARGS"
-    value: "--prefix={{ jenkins_url_prefix }}{{ ('Debian' in ansible_os_family) | ternary(' --httpsPort={{ jenkins_https_port }} --httpsKeyStore={{ jenkins_jks }} --httpsKeyStorePassword={{ jenkins_jks_pass }}','')}}"
+    value: "--prefix={{ jenkins_url_prefix }}{{ ('Debian' in ansible_os_family) | ternary(' --httpPort={{ jenkins_http_port }}','')}}"
```
Comment (dasJ) [inline defaults/main.yml]: Maybe allow binding to an IP?

File: meta/main.yml [hunk 24]
```diff
@@ -1,6 +1,5 @@
...
 
 galaxy_info:
```
Comment (dasJ) [inline meta/main.yml]: Aren't you the author now?

File: tasks/main.yml (hunk 6 above)
Comment (dasJ) [inline tasks/main.yml]: And also install curl

File: tasks/main.yml (hunk 7 above)
Comment (dasJ) [inline tasks/main.yml]: `state: present` is not required because it's the  default

File: tasks/main.yml [hunk 25]
```diff
@@ -30,7 +44,23 @@
...
+    enabled: yes
+
+# Set "Use Jenkins CLI" to true
+- name: Enable Jenkins CLI
+  template:
+    src: jenkins.CLI.xml.j2
+    dest: "{{ jenkins_home }}/jenkins.CLI.xml"
+
+# Add admin account
+- name: configure default users
```
Comment (dasJ) [inline tasks/main.yml]: Capitalize

File: tasks/main.yml (hunk 12 above)
Comment (dasJ) [inline tasks/main.yml]: @SlothOfAnarchy FYI, the Java SSH implementation does not support ED25519 because it's too new (2014) for Java developers

File: tasks/main.yml (hunk 13 above)
Comment (dasJ) [inline tasks/main.yml]: @SlothOfAnarchy Why would there be one?

Comment (dasJ) [inline tasks/main.yml]: When having a jenkins_port, the resulting URL is `http://hostname8080` without the colon required

File: templates/basic-security.groovy.j2 (hunk 14 above)
Comment (dasJ) [inline templates/basic-security.groovy.j2]: Also, this should probably not be the first line because groovy scripts seem to require `#!groovy` as first line. Just add it as second line

File: templates/basic-security.groovy.j2 [hunk 26]
```diff
@@ -1,13 +1,10 @@
...
-#!groovy
 import hudson.security.*
 import jenkins.model.*
 
 def instance = Jenkins.getInstance()
 
 println "--> Checking if security has been set already"
 
-if (!instance.isUseSecurity()) {
     println "--> creating local user 'admin'"
```
Comment (dasJ) [inline templates/basic-security.groovy.j2]: Please redindent this part

File: templates/config.xml.j2 [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+    <extensions/>
+  </scm>
+  <canRoam>true</canRoam>
+  <disabled>false</disabled>
+  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
+  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
+  <triggers>
+    {% if jenkins_timer_trigger is defined %}
+    <hudson.triggers.TimerTrigger>
+      <spec>{{jenkins_timer_trigger}}</spec>
```
Comment (dasJ) [inline templates/config.xml.j2]: What about multiple timers?

File: templates/credentials.xml (hunk 21 above)
Comment (dasJ) [inline templates/credentials.xml]: It is, but that's okay in this case. Usually it's generated when creating the credentials in the UI.

File: templates/credentials.xml [hunk 28]
```diff
@@ -0,0 +1,19 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<com.cloudbees.plugins.credentials.SystemCredentialsProvider plugin="credentials@2.1.13">
```
Comment (dasJ) [inline templates/credentials.xml]: @SlothOfAnarchy You missed this hard coded version

Comment (SlothOfAnarchy) [inline templates/credentials.xml]: No, I just thought I mentioned it often enough so that my concern in general would be clear.
